src/lib.rs

```diff
@@ -59,18 +59,22 @@ pub trait TcpStack {
 /// module. You could have another implemenation which knows how to driver the
 /// Rust Standard Library's `std::net` module. Given this trait, you can how
 /// write a portable CoAP client which can work with either implementation.
-pub trait UdpStack {
+pub trait UdpClient {
 	/// The type returned when we create a new UDP socket
 	type UdpSocket;
 	/// The type returned when we have an error
 	type Error: core::fmt::Debug;
 
-	/// Open a new UDP socket to the given address and port. UDP is connectionless,
-	/// so unlike `TcpStack` no `connect()` is required.
-	fn open(&self, remote: SocketAddr) -> Result<Self::UdpSocket, Self::Error>;
+	/// Open a UDP socket with a dynamically selected port.
+	///
+	/// Selects a port number automatically and initializes for read/writing.
+	fn connect(&mut self, remote: SocketAddr) -> Result<Self::UdpSocket, Self::Error>;
 
 	/// Send a datagram to the remote host.
-	fn write(&self, socket: &mut Self::UdpSocket, buffer: &[u8]) -> nb::Result<(), Self::Error>;
+	///
+	/// The remote host used is either the one specified in `UdpStack::connect`
+	/// or the last one used in `UdpServerStack::write_to`.
+	fn send(&mut self, socket: &mut Self::UdpSocket, buffer: &[u8]) -> nb::Result<(), Self::Error>;
```

Comment:
There [was a reason](https://github.com/rust-embedded-community/embedded-nal/issues/4) that none of these used a mutable reference

Reply:
Hmm, I'm not sure how to get my impl to work without that.  Here's my send function for example:

```rust
    fn send(&mut self, socket: &mut Self::UdpSocket, buffer: &[u8]) -> nb::Result<(), Self::Error> {
        socket.send(&mut self.bus, buffer)?;
        Ok(())
    }
```

Is there a way to internalize the need for a mutable borrow of the `bus`?

Reply:
nvm, found an article about it. https://ricardomartins.cc/2016/06/08/interior-mutability